reco_utils/dataset/amazon_reviews.py

```diff
@@ -233,6 +242,87 @@ def _data_generating(input_file, train_file, valid_file, test_file, min_sequence
+    '''produce train, valid and test file from processed_output file
+    Each user's behavior sequence will only produce one line in trian file.
+    Like, user's behavior sequence: 12345, and this function will write into train file: 12345
+    '''
+    f_input = open(input_file, "r")
+    f_train = open(train_file, "w")
+    f_valid = open(valid_file, "w")
+    f_test = open(test_file, "w")
+    print("data generating...")
+
+    last_user_id = None
+    last_movie_id = None
+    last_category = None
+    last_datetime = None
+    last_tfile = None
+    for line in f_input:
+        line_split = line.strip().split("\t")
+        tfile = line_split[0]
+        label = int(line_split[1])
+        user_id = line_split[2]
+        movie_id = line_split[3]
+        date_time = line_split[4]
+        category = line_split[5]
+
+        if last_tfile == "train":
+            fo = f_train
+        elif last_tfile == "valid":
+            fo = f_valid
+        elif last_tfile == "test":
+            fo = f_test
+        if user_id != last_user_id or tfile == "valid" or tfile == "test":
+            if last_user_id != None:
+                history_clk_num = len(movie_id_list)
+                cat_str = ""
+                mid_str = ""
+                dt_str = ""
+                for c1 in cate_list[:-1]:
+                    cat_str += c1 + ","
+                for mid in movie_id_list[:-1]:
+                    mid_str += mid + ","
+                for dt_time in dt_list[:-1]:
+                    dt_str += dt_time + ","
+                if len(cat_str) > 0:
+                    cat_str = cat_str[:-1]
+                if len(mid_str) > 0:
+                    mid_str = mid_str[:-1]
+                if len(dt_str) > 0:
+                    dt_str = dt_str[:-1]
+                if history_clk_num > min_sequence:
+                    fo.write(
+                        line_split[1]
+                        + "\t"
+                        + last_user_id
+                        + "\t"
+                        + last_movie_id
+                        + "\t"
+                        + last_category
+                        + "\t"
+                        + last_datetime
+                        + "\t"
+                        + mid_str
+                        + "\t"
+                        + cat_str
+                        + "\t"
+                        + dt_str
+                        + "\n"
+                    )
+            if tfile == "train" or last_user_id == None:        
+                movie_id_list = []
+                cate_list = []
+                dt_list = []
+        last_user_id = user_id
+        last_movie_id = movie_id
+        last_category = category
+        last_datetime = date_time
+        last_tfile = tfile
+        if label:
+            movie_id_list.append(movie_id)
+            cate_list.append(category)
+            dt_list.append(date_time)
 
 def _create_item2cate(instance_file):
```

Comment:
What's the reason to use global variable here instead of returning `item2cate`?

Reply:
mmm, I don't see a justification to use a global here, there are multiple problems in this function, created an issue to fix them on a different pr https://github.com/microsoft/recommenders/issues/1141